### DISCUSSION TITLE
Can getters be reactive to x-data changes?

### DISCUSSION THREAD

**Discussion by kslstn** [alpinejs/alpine#3305] [1. Help] [answered]
I tried creating a getter that returns a value based on other x-data variables. Like this:

```
get myGetter(){
  this.a;
  return this.a + 2;
}
```

As you can see, I've already added the `this.a;` line hoping Alpine would reevaluate the getter once `a` changes. This doesn't work though—the getter just returns a result based on `a`'s initial value. A `console.log()` shows that the never gets reevaluated. What does work is a simple method:

```
myMethod(){
  return this.a + 2;
}
```

[Here's a Codepen](https://codepen.io/kslstn/pen/mdKzEdP).

Is it better to not use getters in Alpine when data can change, or is there a better way to define them?

**Comment by SimoTod** [comment, marked as answer]
Just
```js
get myGetter(){
  return this.a + 2;
}
```
this is already rebound to be the proxied context.

It's not working because you are doing something weird with your x-data. Try `x-data="myApp"`

**Reply by kslstn** [reply]
Oh you're right! That weird something with my x-data is a spread operator, which allows one to bind multiple encapsulated parts. For things like reusable methods. I thought that was the recommended way to do that in Alpine 3, but perhaps there's a better way for that in Alpine 3?

**Reply by SimoTod** [reply]
I think it's a popular way but I wouldn't say recommend as it's never mentioned in the docs.

Properties need to be modified through the proxy layer to be reactive, `this` in getters is the real object by default but it gets rebound by alpine and becomes the proxy. Using the spread operator, though, seems to mess up with the `this` context.

**Reply by ekwoka** [reply]
You can't spread getters. It will just use the value at that moment.

If you want a utility component that merges other components, you need to use 

Object.defineProperties

And

Object.getPropertyDescriptors

Getting the descriptors will include the getters and setters, and such and allow you to transfer them to a new object with rebound `this`.

**Reply by ekwoka** [reply]
As an example of above you could do

```js
Alpine.data('mergeComponents', (...components) => components.reduce(
  (acc, next) => {
    Object.defineProperties(
      acc,
      Object.getOwnPropertyDescriptors(typeof next === 'function' ? next() : next)
    )
  }, {}
)
```


Then this can be used to combine component together

**Reply by kslstn** [reply]
I noticed that properties aren't reevaluated when their based on arrays.

I made a little test to figure out what was going on:
```js
		fruit: ['Apple', 'Banana', 'Orange'],
		visibleFruit: ['Melon'],
		someProperty: false,
		
		addFruit(){
			this.visibleFruit.push(this.fruit[Math.floor(Math.random() * 3)]);
			this.someProperty = !this.someProperty
		},
		
		removeFruit(){
			this.visibleFruit.pop();
		},
		
		get lastFruit(){
			this.someProperty;
			return document.querySelector('li:last-child').innerHTML;
		},
```
`lastfruit()` is only evaluated when I change `someProperty`, which I only added to trigger that. When I only push/pop/reassign the `visibleFruit` array like in `removeFruit()`, `lastFruit` stays the same. Adding `someProperty` is obviously not pretty. I'm surprised that array changes don't trigger an evaluation of `lastFruit`. Am I doing something wrong? Shouldn't this be mentioned in the docs [section about getters](https://alpinejs.dev/directives/data#getters)?

Check [this Codepen](https://codepen.io/kslstn/pen/NWBWKxP?editors=1111) for the test above.

**Reply by SimoTod** [reply]
You never touch visibleFruit in your getter so it's expected.

**Reply by kslstn** [reply]
Ah sorry, forgot to put that back into the Codepen. Even with `visibleFruit` it's not working:
```
		get lastFruit(){
			this.someProperty;
			thits.visibleFruit;
			return document.querySelector('li:last-child').innerHTML;
		},
```

**Reply by SimoTod** [reply]
To clarify, yeah you are actually approaching the problem in the wrong way. You should always use the state. The dom is just the final rappresentation you shouldn't really use element.innerHtml to read the value of the last element of your array

**Reply by SimoTod** [reply]
@kslstn https://codepen.io/SimoTod/pen/vYaYERR?editors=1111

**Reply by kslstn** [reply]
Thanks, it's just that the reason I'm doing it this way, is that I need to get the last item's `scrollTop` and `scrollLeft`.  I'm fine with getting the element some other way though, but I can't use `x-ref` here.

**Reply by SimoTod** [reply]
The reactivity is never deep, it only triggers when a property mentioned in the code changes (aka is reassigned). Changing a subproperty (or an array element in your case) never triggers the reactivity. In your case, what changes is the length of the array so touch that in your function.

**Reply by kslstn** [reply]
Now a lot of things start to make sense! Thanks!!
